### PR TITLE
[Issue 12812] Introduce ZKBatchedMetadataStore.

### LINKED ISSUE #12812: Optimize zookeeper client performance for loading amounts of topics.
**Is your enhancement request related to a problem? Please describe.**

Optimize zookeeper client performance for loading amounts of topics.
Our use case is described here  #12651 .

**Describe the solution you'd like**

Introduce [zk multi ops](http://zookeeper.apache.org/doc/r3.7.0/apidocs/zookeeper-server/org/apache/zookeeper/ZooKeeper.html#multi-java.lang.Iterable-org.apache.zookeeper.AsyncCallback.MultiCallback-java.lang.Object-) to optimize the zk client performance.

Our team did a perf test on the zk multi ops, here is the result.

|Op type | Single Op (ops/s) | Batch 8 with Multi (ops/s)|
|--|--|--|
|Create|25k|60k|
|Write(with pre-created 5M nodes)|25|90k|
|Read|25k|200k|

Here is the settings:
- Single node data size : 256 Bytes
- ZK server version: 3.6.3
- ZK cluster nodes: 3
- ZK server disk: 8TB Nvme SSD
- Cpu usage: 6/40 on core bare metal server.
- OS : centos 7

It's clear that with multi, we can achieve much more performance with the same cpu usage, especially with read operations.

The basic idea of implementation will be add two queue (one for read ops and one for write ops) in PulsarZooKeeperClient, all zk ops will be added to the queue first, and  a background thread will batch theses requests and sends to zk server in one "multi op".
This implementation will introduce the follow parameter in broker configs:
- **enableAutoBatchZookeeperOps**, this feature is optional, as it may increase metadata latency with small amount of topics.
- **autoBatchZookeeperOpsMaxNum** and **autoBatchZookeeperOpsMaxDelayMills**  Just like auto batching parameters in pulsar producer. Limits the max number of ops in one batch and max delay time  to wait for a batch.

**Describe alternatives you've considered**
Add a ratelimiter for topic loading, see  #12651 

**Additional context**
Here is the explanation why multi ops works much better with read ops. 
- All write ops needs to go through a single thread processor in master.  And we have reach the max qps in the previous perf test.
- Read ops can be handled by slave nodes. We can get more qps with larger batch size in the previous perf test.

Here is the flame graph of the bottleneck thread in master.
image

### PR DESCRIPTION
**Pull request by Jason918**
Fixes #12812

### Motivation

See #12812

### Modifications

Add class `ZKBatchedMetadataStore` extends `ZKMetadataStore` and overrides `storeGet` and `getChildrenFromStore`.
Put these two kinds of read operations into `org.apache.pulsar.metadata.impl.batch.BatchWorker#opQueue`,  which contains a background thread that will send batch ops once it reaches target op number (set as 100) or waited long enough (5ms).

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change is already covered by existing tests, all test cases in pulsar-metadata.

This change added tests and can be verified as follows:
  - ZKBatchedMetadataStoreTest

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

Check the box below and label this PR (if you have committer privilege).

Need to update docs? 
  
- [x] `no-need-doc` 

Pref optimization.

### CODE REVIEW

**Review by merlimat — COMMENTED** [review]
This is implementing the batching for `getData`/`getChildren` calls but not for the `create` and `setData` which will be very interesting, even though we need to solve the problem of creating the parent nodes. 

I'll submit a different PR with the logic for the write/create

**Comment by Jason918** [comment]
> This is implementing the batching for `getData`/`getChildren` calls but not for the `create` and `setData` which will be very interesting, even though we need to solve the problem of creating the parent nodes.
> 
> I'll submit a different PR with the logic for the write/create

This is WIP. Adding create and setData is pretty easy based on this PR.

I plan to get a comparison result of containing batch read operations first, as we can see in previous zk perf result #12812, batch reading have the most performance improvements.

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/ZKBatchedMetadataStore.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+import org.apache.pulsar.metadata.impl.batch.BatchWorker;
+import org.apache.pulsar.metadata.impl.batch.MetadataOp;
+import org.apache.zookeeper.AsyncCallback;
+import org.apache.zookeeper.KeeperException.Code;
+import org.apache.zookeeper.Op;
+import org.apache.zookeeper.OpResult;
+import org.apache.zookeeper.ZooDefs;
+
+@Slf4j
+public class ZKBatchedMetadataStore extends ZKMetadataStore implements AsyncCallback.MultiCallback {
```
[the remark this answers is not in the document]
Comment (Jason918) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/ZKBatchedMetadataStore.java:41]: I wrote a AbstractBatchedMetadataStore at first, but got some issue. 
- We have to use two different batching queues for Zookeeper Multi operations, one for read and one for write.  But I believe that we don't need to do that in other metadata store implementations, e.g. Memory, RocksDB or maybe etcd version of implementation.
- There are not many things left in this AbstractBatchedMetadataStore, mostly is in wrapped in `BatchWorker`, maybe only `fallbackToSingleOps`.

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/ZKBatchedMetadataStore.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+    protected final boolean metadataStoreBatchingEnable = true;
+    protected final int metadataStoreBatchingMaxDelayMillis = 5;
+    protected final int metadataStoreBatchingMaxOperations = 100;
+    protected final int metadataStoreBatchingMaxSizeKb = 128;
```
[the remark this answers is not in the document]
Comment (Jason918) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/ZKBatchedMetadataStore.java:46]: OK, I will fix this.
@codelipenghui  @hangc0276 FYI.

**Comment by merlimat** [comment]
@Jason918 I've added a different PR based on the approach I was mentioning: #13043. PTAL

**Comment by Jason918** [comment]
> @Jason918 I've added a different PR based on the approach I was mentioning: #13043. PTAL

OK. This seems to be a duplicated work. Let's continue this work on your PR.

**Pull request closed without merging**